File: frontend/src/utils/homeDeviceHealth.ts
/** Home dashboard health display (PAS-52). */

import { DEVICE_LABEL } from './terminology'

/** CPU/memory from the health snapshot. Never GPU occupancy (gpu-devices). */
export function deviceResourcesLine(device: {
  reachability?: string
  resources?: {
    cpuLoadPercent?: number | null
    memoryUsedMB?: number | null
    memoryTotalMB?: number | null
  } | null
}): string | null {
  if (device.reachability !== 'ok' || !device.resources) return null
  const parts: string[] = []
  if (device.resources.cpuLoadPercent != null) {
    parts.push(`CPU ${Math.round(device.resources.cpuLoadPercent)}%`)
  }
  if (device.resources.memoryUsedMB != null && device.resources.memoryTotalMB != null) {
    parts.push(
      `${(device.resources.memoryUsedMB / 1024).toFixed(1)} / ${(device.resources.memoryTotalMB / 1024).toFixed(0)} GB`,
    )
  }
  return parts.length ? parts.join(' · ') : null
}

export function deviceWorkloadLine(device: {
  reachability?: string
  workloadCount?: number | null
  healthCounts?: Record<string, number> | null
}): string {
  if (device.reachability !== 'ok') return 'Health unavailable'
  if (device.workloadCount == null) return 'Workloads unknown'
  const count = device.workloadCount
  const failed = device.healthCounts?.failed ?? 0
  if (failed > 0) return `${count} workloads · ${failed} failed`
  return `${count} workload${count === 1 ? '' : 's'}`
}

/** A successful summarize() always includes health keys (even all zeros).
 *  Empty `{}` means no device returned a health summary. */
export function hasKnownHealthCounts(
  counts?: Record<string, number> | null,
): counts is Record<string, number> {
  return counts != null && Object.keys(counts).length > 0
}

export function resolveHealthCounts(
  preferred?: Record<string, number> | null,
  fallback?: Record<string, number> | null,
): Record<string, number> {
  if (hasKnownHealthCounts(preferred)) return preferred
  return fallback ?? {}
}

/** Home-wide "X of Y workloads running". Null when the API reports unknown. */
export function homeWorkloadsRunningLine(
  totals: { workloadCount?: number | null } | null | undefined,
  runningCount: number,
): string | null {
  if (totals?.workloadCount == null) return null
  return `${runningCount} of ${totals.workloadCount} workloads running`
}

export function isReachabilityOk(code: string | undefined): boolean {
  return code === 'ok'
}

/** Pill on Device cards and Device detail. `memberHTTP` is not Unreachable. */
export function reachabilityLabel(code: string | undefined): string {
  switch (code) {
    case 'ok':
      return 'Reachable'
    case 'connectTimeout':
      return 'Timed out'
    case 'cancelled':
      return 'Cancelled'
    case 'tlsFailure':
      return 'TLS failed'
    case 'memberHTTP':
      return 'HTTP error'
    default:
      return 'Unreachable'
  }
}

export function reachabilityPillClass(code: string | undefined): string {
  if (code === 'ok') return 'running'
  if (code === 'memberHTTP') return 'degraded'
  return 'failed'
}

export function reachabilityCardClass(code: string | undefined): string {
  if (code === 'ok' || !code) return ''
  if (code === 'memberHTTP') return 'http-error'
  return 'unreachable'
}

export function reachabilityHint(
  device: { reachability?: string; reachabilityError?: string | null },
): string | null {
  const code = device.reachability
  if (!code || code === 'ok') return null
  if (code === 'unreachable') {
    return `This ${DEVICE_LABEL.toLowerCase()} is still running locally. The member did not answer.`
  }
  const error = device.reachabilityError?.trim()
  return error || reachabilityLabel(code)
}

File: Sources/BarkVisorCore/Models/HomeDeviceHealth.swift
import Foundation

/// Live facts for one Device on the Home dashboard (PAS-52).
///
/// Missing resource / health fields mean the member was unreachable or the
/// probe returned no body. Listing still includes the Device.
public struct HomeDeviceLiveFacts: Sendable, Equatable {
    public var displayName: String?
    public var collectedAt: String?
    public var platform: HomeDevicePlatformSummary?
    public var resources: HomeDeviceResourceSummary?
    public var workloadCount: Int?
    public var healthCounts: [String: Int]?

    public init(
        displayName: String? = nil,
        collectedAt: String? = nil,
        platform: HomeDevicePlatformSummary? = nil,
        resources: HomeDeviceResourceSummary? = nil,
        workloadCount: Int? = nil,
        healthCounts: [String: Int]? = nil,
    ) {
        self.displayName = displayName
        self.collectedAt = collectedAt
        self.platform = platform
        self.resources = resources
        self.workloadCount = workloadCount
        self.healthCounts = healthCounts
    }
}

public struct HomeDevicePlatformSummary: Codable, Sendable, Equatable {
    public var os: String
    public var arch: String

    public init(os: String, arch: String) {
        self.os = os
        self.arch = arch
    }
}

public struct HomeDeviceResourceSummary: Codable, Sendable, Equatable {
    public var cpuCount: Int?
    public var memoryTotalMB: Int?
    public var memoryUsedMB: Int?
    public var cpuLoadPercent: Double?

    public init(
        cpuCount: Int? = nil,
        memoryTotalMB: Int? = nil,
        memoryUsedMB: Int? = nil,
        cpuLoadPercent: Double? = nil,
    ) {
        self.cpuCount = cpuCount
        self.memoryTotalMB = memoryTotalMB
        self.memoryUsedMB = memoryUsedMB
        self.cpuLoadPercent = cpuLoadPercent
    }
}

/// One row in `GET /api/home/devices/health`.
public struct HomeDeviceHealthSnapshot: Codable, Sendable, Equatable {
    public var hostId: String
    public var role: String
    public var displayName: String?
    public var fingerprint: String?
    public var agentHost: String?
    public var agentPort: Int
    public var pairedAt: String?
    public var reachability: String
    public var reachabilityError: String?
    public var collectedAt: String?
    public var platform: HomeDevicePlatformSummary?
    public var resources: HomeDeviceResourceSummary?
    public var workloadCount: Int?
    public var healthCounts: [String: Int]?

    public init(
        hostId: String,
        role: String,
        displayName: String? = nil,
        fingerprint: String? = nil,
        agentHost: String? = nil,
        agentPort: Int = Config.agentPort,
        pairedAt: String? = nil,
        reachability: String,
        reachabilityError: String? = nil,
        collectedAt: String? = nil,
        platform: HomeDevicePlatformSummary? = nil,
        resources: HomeDeviceResourceSummary? = nil,
        workloadCount: Int? = nil,
        healthCounts: [String: Int]? = nil,
    ) {
        self.hostId = hostId
        self.role = role
        self.displayName = displayName
        self.fingerprint = fingerprint
        self.agentHost = agentHost
        self.agentPort = agentPort
        self.pairedAt = pairedAt
        self.reachability = reachability
        self.reachabilityError = reachabilityError
        self.collectedAt = collectedAt
        self.platform = platform
        self.resources = resources
        self.workloadCount = workloadCount
        self.healthCounts = healthCounts
    }
}

public struct HomeDeviceHealthTotals: Codable, Sendable, Equatable {
    public var devices: Int
    public var reachable: Int
    public var unreachable: Int
    /// Sum of known per-device counts. `nil` when any reachable Device
    /// did not return a workload health summary — never treat unknown as 0.
    public var workloadCount: Int?
    public var healthCounts: [String: Int]

    public init(
        devices: Int,
        reachable: Int,
        unreachable: Int,
        workloadCount: Int?,
        healthCounts: [String: Int],
    ) {
        self.devices = devices
        self.reachable = reachable
        self.unreachable = unreachable
        self.workloadCount = workloadCount
        self.healthCounts = healthCounts
    }
}

/// `GET /api/home/devices/health` body.
public struct HomeDeviceHealthReport: Codable, Sendable, Equatable {
    public var devices: [HomeDeviceHealthSnapshot]
    public var totals: HomeDeviceHealthTotals

    public init(devices: [HomeDeviceHealthSnapshot], totals: HomeDeviceHealthTotals) {
        self.devices = devices
        self.totals = totals
    }
}
